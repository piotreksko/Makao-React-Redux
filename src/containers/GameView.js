import React, { Component } from "react";
import { connect } from "react-redux";
import * as logicActions from "../actions/logicActions";
import * as soundActions from "../actions/soundActions";
import Aux from "../hoc/Auxilliary";
import Modals from "./Modals";
import Header from "./Header";
import Player from "./Player";
import Icons from "./Icons";
import Deck from "../components/Deck";
import Pile from "../components/Pile";
import Confetti from "react-dom-confetti";
import CpuPlayer from "../components/CpuPlayer";

export class GameView extends Component {
  constructor(props) {
    super(props);

    this.takeCards = this.takeCards.bind(this);
    this.restartGame = this.restartGame.bind(this);
  }

  componentWillMount() {
    setTimeout(() => {
      this.props.showModal("whoStarts");
    }, 1);
    setTimeout(() => {
      this.props.hideModal("whoStarts");
    }, 1000);
  }

  takeCards() {
    this.props.takeCards(this.props.gameState.cardsToTake);
    this.props.playSound("take_card");
    this.props.endTurn();
  }

  restartGame() {
    this.props.restartGame();
    this.props.hideModal("gameOver");
  }

  render() {
    const gameState = this.props.gameState,
    playerCanMove = !gameState.player.wait && !gameState.waitTurn ? true : false,
    { pile } = gameState;

    const confettiConfig = {
      angle: 90,
      spread: 100,
      startVelocity: 50,
      elementCount: 100,
      decay: 0.9
    };

    return (
      <Aux>
        <Modals restartGame={this.restartGame} />
        <div className="confetti">
          <Confetti
            active={!this.props.gameState.player.cards.length}
            config={confettiConfig}
          />
        </div>
        <Header />
        <CpuPlayer cpuPlayer={gameState.cpuPlayer} />
        <div className="flex-container middle">
          <Deck takeCard={this.takeCards} playerCanMove={playerCanMove} />
          <Pile cards={pile} />
          <Icons />
        </div>
        <Player />
      </Aux>
    );
  }
}

const mapStateToProps = state => {
  return {
    globalStats: state.stats.global,
    currentStats: state.stats.current,
    modals: state.modals,
    gameState: state.gameState
  };
};

const mapDispatchToProps = dispatch => {
  return {
    showModal: modal => dispatch({ type: "SHOW_MODAL", modal: modal }),
    hideModal: modal => dispatch({ type: "HIDE_MODAL", modal: modal }),
    takeCards: howMany => dispatch(logicActions.takeCards("player")),
    updateGameFactor: (factor, value) =>
      dispatch(logicActions.updateGameFactor(factor, value)),
    endTurn: () => {
      dispatch(logicActions.updateGameFactor("playerTurn", false));
      dispatch(logicActions.makeCpuMove());
      dispatch(logicActions.checkMacaoAndWin());
    },
    restartGame: () => dispatch({ type: logicActions.RESTART_GAME }),
    playSound: soundName => dispatch(soundActions.playSound(soundName))
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(GameView);
