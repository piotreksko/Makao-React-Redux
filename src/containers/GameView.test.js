import React from 'react';
import {shallow} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import GameView from './GameView';

const mockStore = configureMockStore([ thunk ]);

const storeStateMock = {
  gameState:{
    player: {
      cards: [],
      wait: 0
    },
    cpuPlayer: {
      cards: [],
      wait: 0
    },
    pile: [],
    deck: [],
    chosenWeight: "",
    chosenType: "",
    isPlayerTurn: 0,
    firstCardChecked: false,
    jackActive: 0,
    cardsToTake: 1,
    waitTurn: 0,
    battleCardActive: false,
    winner: 0,
    gameOver: false
  }
};

let store;
let component;
describe('tests for GameView container', () => {
  beforeEach(() => {
    store = mockStore(storeStateMock);
    component = shallow(<GameView store={store} />).shallow();
  });

  it('renders properly', () => {
    expect(component).toMatchSnapshot();
  });
});