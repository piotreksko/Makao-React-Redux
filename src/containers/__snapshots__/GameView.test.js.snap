// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for GameView container renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GameView
    endTurn={[Function]}
    gameState={
      Object {
        "battleCardActive": false,
        "cardsToTake": 1,
        "chosenType": "",
        "chosenWeight": "",
        "cpuPlayer": Object {
          "cards": Array [],
          "wait": 0,
        },
        "deck": Array [],
        "firstCardChecked": false,
        "gameOver": false,
        "isPlayerTurn": 0,
        "jackActive": 0,
        "pile": Array [],
        "player": Object {
          "cards": Array [],
          "wait": 0,
        },
        "waitTurn": 0,
        "winner": 0,
      }
    }
    hideModal={[Function]}
    makePlayerMove={[Function]}
    playSound={[Function]}
    restartGame={[Function]}
    showModal={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
    takeCards={[Function]}
    updateGameFactor={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Connect(Modals)
          restartGame={[Function]}
        />,
        <div
          className="confetti"
        >
          <Confetti
            active={true}
            config={
              Object {
                "angle": 90,
                "decay": 0.9,
                "elementCount": 30,
                "spread": 80,
                "startVelocity": 50,
              }
            }
          />
        </div>,
        <Connect(Header)
          restartGame={[Function]}
        />,
        <CpuPlayer
          cpuPlayer={
            Object {
              "cards": Array [],
              "wait": 0,
            }
          }
        />,
        <div
          className="flex-container middle cards-container"
        >
          <Deck
            cardsInDeck={0}
            firstCardChecked={false}
            isPlayerTurn={0}
            playerCanMove={true}
            takeCard={[Function]}
          />
          <Pile
            cards={Array []}
          />
          <Connect(Icons) />
        </div>,
        <Connect(Player) />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "restartGame": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Confetti
            active={true}
            config={
              Object {
                "angle": 90,
                "decay": 0.9,
                "elementCount": 30,
                "spread": 80,
                "startVelocity": 50,
              }
            }
          />,
          "className": "confetti",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": true,
            "config": Object {
              "angle": 90,
              "decay": 0.9,
              "elementCount": 30,
              "spread": 80,
              "startVelocity": 50,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "restartGame": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "cpuPlayer": Object {
            "cards": Array [],
            "wait": 0,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Deck
              cardsInDeck={0}
              firstCardChecked={false}
              isPlayerTurn={0}
              playerCanMove={true}
              takeCard={[Function]}
            />,
            <Pile
              cards={Array []}
            />,
            <Connect(Icons) />,
          ],
          "className": "flex-container middle cards-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "cardsInDeck": 0,
              "firstCardChecked": false,
              "isPlayerTurn": 0,
              "playerCanMove": true,
              "takeCard": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "cards": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Connect(Modals)
            restartGame={[Function]}
          />,
          <div
            className="confetti"
          >
            <Confetti
              active={true}
              config={
                Object {
                  "angle": 90,
                  "decay": 0.9,
                  "elementCount": 30,
                  "spread": 80,
                  "startVelocity": 50,
                }
              }
            />
          </div>,
          <Connect(Header)
            restartGame={[Function]}
          />,
          <CpuPlayer
            cpuPlayer={
              Object {
                "cards": Array [],
                "wait": 0,
              }
            }
          />,
          <div
            className="flex-container middle cards-container"
          >
            <Deck
              cardsInDeck={0}
              firstCardChecked={false}
              isPlayerTurn={0}
              playerCanMove={true}
              takeCard={[Function]}
            />
            <Pile
              cards={Array []}
            />
            <Connect(Icons) />
          </div>,
          <Connect(Player) />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "restartGame": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Confetti
              active={true}
              config={
                Object {
                  "angle": 90,
                  "decay": 0.9,
                  "elementCount": 30,
                  "spread": 80,
                  "startVelocity": 50,
                }
              }
            />,
            "className": "confetti",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "active": true,
              "config": Object {
                "angle": 90,
                "decay": 0.9,
                "elementCount": 30,
                "spread": 80,
                "startVelocity": 50,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "restartGame": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "cpuPlayer": Object {
              "cards": Array [],
              "wait": 0,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Deck
                cardsInDeck={0}
                firstCardChecked={false}
                isPlayerTurn={0}
                playerCanMove={true}
                takeCard={[Function]}
              />,
              <Pile
                cards={Array []}
              />,
              <Connect(Icons) />,
            ],
            "className": "flex-container middle cards-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "cardsInDeck": 0,
                "firstCardChecked": false,
                "isPlayerTurn": 0,
                "playerCanMove": true,
                "takeCard": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "cards": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
